### PR TITLE
res_pjsip_pubsub: Add new pubsub module capabilities.

### PR DESCRIPTION
**Pull request by InterLinked1**
The existing res_pjsip_pubsub APIs are somewhat limited in what they can do. This adds a few API extensions that make it possible for PJSIP pubsub modules to implement richer features than is currently possible.

* Allow pubsub modules to get a handle to pjsip_rx_data on subscription
* Allow pubsub modules to run a callback when a subscription is renewed
* Allow pubsub modules to run a callback for outgoing NOTIFYs, with a handle to the tdata, so that modules can append their own headers to the NOTIFYs

This change does not add any features directly, but makes possible several new features that will be added in future changes.

Resolves: #81
ASTERISK-30485 #close
Imported from Gerrit: https://gerrit.asterisk.org/c/asterisk/+/20031

Master-Only: True

### CODE REVIEW

File: include/asterisk/res_pjsip_pubsub.h [hunk 1]
```diff
@@ -282,6 +294,25 @@ struct ast_sip_notifier {
...
+	/*!
+	 * \brief Called when a SUBSCRIBE arrives for an already active subscription.
+	 *
+	 * \param sub The existing subscription
+	 * \retval 0 Success
+	 * \retval -1 Failure
+	 */
+	int (*refresh_subscribe)(struct ast_sip_subscription *sub, pjsip_rx_data *rdata);
+	/*!
+	 * \brief Optional callback to execute before sending outgoing NOTIFY responses.
```
Comment (jcolp) [inline include/asterisk/res_pjsip_pubsub.h:306]: A NOTIFY is a request, not a response

File: include/asterisk/res_pjsip_pubsub.h [hunk 2]
```diff
@@ -282,6 +294,25 @@ struct ast_sip_notifier {
...
+	 *        Because res_pjsip_pubsub creates the tdata internally, this allows modules
+	 *        to access the tdata if needed, e.g. to add custom headers.
+	 *
+	 * \param sub The existing subscription
+	 * \param tdata The pjsip_tx_data to use for the outgoing NOTIFY
+	 * \retval 0 Success
+	 * \retval -1 Failure
+	 */
+	int (*tdata_callback)(struct ast_sip_subscription *sub, pjsip_tx_data *tdata);
```
Comment (jcolp) [inline include/asterisk/res_pjsip_pubsub.h:315]: tdata_callback is not a very descriptive name in comparison to the others, something indicating the callback is before the NOTIFY is sent would be better such as:

notify_created

File: res/res_pjsip_pubsub.c [hunk 3]
```diff
@@ -807,6 +807,30 @@ static struct ast_sip_subscription_handler *subscription_get_handler_from_rdata(
...
+/*! \brief Retrieve a handler using the Event header of a tdata message */
+static struct ast_sip_subscription_handler *subscription_get_handler_from_tdata(pjsip_tx_data *tdata, const char *endpoint)
+{
+	pjsip_event_hdr *event_header;
+	char event[32];
+	struct ast_sip_subscription_handler *handler;
+
+	event_header = pjsip_msg_find_hdr_by_name(tdata->msg, &str_event_name, tdata->msg->hdr.next);
+	if (!event_header) {
+		ast_log(LOG_WARNING, "Incoming SUBSCRIBE request from %s with no Event header\n",
```
Comment (jcolp) [inline res/res_pjsip_pubsub.c:819]: This isn't an accurate warning message.

File: res/res_pjsip_pubsub.c [hunk 4]
```diff
@@ -2491,6 +2529,13 @@ static int send_notify(struct sip_subscription_tree *sub_tree, unsigned int forc
 		pjsip_msg_add_hdr(tdata->msg, (pjsip_hdr *) require);
 	}
 
+	handler = subscription_get_handler_from_tdata(tdata, sub_tree->persistence->endpoint);
```
Comment (jcolp) [inline res/res_pjsip_pubsub.c:2532]: The handler should already be accessible directly at:

sub_tree->root->handler

And guaranteed to be non-NULL

File: res/res_pjsip_pubsub.c [hunk 5]
```diff
@@ -2954,6 +2999,7 @@ static int generate_initial_notify(struct ast_sip_subscription *sub)
 
 	notify_data = sub->handler->notifier->get_notify_data(sub);
 	if (!notify_data) {
+		ast_debug(3, "No notify data, terminating\n");
```
Comment (jcolp) [inline res/res_pjsip_pubsub.c:3002]: I don't think this debug message contains enough information if it is to remain

File: res/res_pjsip_pubsub.c [hunk 6]
```diff
@@ -4085,25 +4124,32 @@ static void pubsub_on_rx_refresh(pjsip_evsub *evsub, pjsip_rx_data *rdata,
 		sub_tree->state = SIP_SUB_TREE_TERMINATE_PENDING;
 	}
 
+	endpoint = ast_pjsip_rdata_get_endpoint(rdata);
```
Comment (jcolp) [inline res/res_pjsip_pubsub.c:4127]: handler also should directly be accessible, and guaranteed to be non-NULL if a refresh

**Review by jcolp — CHANGES_REQUESTED** [review]

**Review by jcolp — DISMISSED** [review]

**Comment by InterLinked1** [comment]
@jcolp Previously the gate tests were failing but this last one, it seems the unit tests also failed. It says `Building failed.  See /github/workspace/cache/output//build.err for more details.` but I can't find where that would be. Is this an actual failure to be concerned about here or just a kink in the CI process still being worked out?

**Comment by jcolp** [comment]
It's a legit build failure:

`   [CC] res_pjsip_sdp_rtp.c -> res_pjsip_sdp_rtp.o
res_pjsip_pubsub.c: In function ‘pubsub_on_rx_refresh’:
res_pjsip_pubsub.c:4102:17: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
 4102 |         handler = sub_tree->root->handler;
      |                 ^
res_pjsip_pubsub.c:4125:52: error: ‘endpoint’ undeclared (first use in this function); did you mean ‘endrpcent’?
 4125 |                         resp = build_resource_tree(endpoint, handler, resource, &tree,
      |                                                    ^~~~~~~~
      |                                                    endrpcent
res_pjsip_pubsub.c:4125:52: note: each undeclared identifier is reported only once for each function it appears in
make[1]: *** [/home/jcolp/development/asterisk/github/Makefile.rules:165: res_pjsip_pubsub.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [Makefile:396: res] Error 2
`

**Comment by jcolp** [comment]
I didn't, I built it locally. I've asked George to comment on the best way to access that.

**Comment by gtjoseph** [comment]
> I didn't, I built it locally. I've asked George to comment on the best way to access that.

It's not as easy as it used to be and it's on my list to fix but in the mean time...

* Click on the details for the failing check.
* Click on the job summary at the top left of the page
* Scroll down to Artifacts and click on Unit Test Output.

It's a zip file that will usually cause the browser to prompt to open or download.  If you click open, many distros will automatically open an archive browser that'll let you read the files inside without having to manually unzip the file first.

I'm going to go back to keeping that output in the initial details.

**Comment by InterLinked1** [comment]
> > I didn't, I built it locally. I've asked George to comment on the best way to access that.
> 
> It's not as easy as it used to be and it's on my list to fix but in the mean time...
> 
> * Click on the details for the failing check.
> * Click on the job summary at the top left of the page
> * Scroll down to Artifacts and click on Unit Test Output.
> 
> It's a zip file that will usually cause the browser to prompt to open or download. If you click open, many distros will automatically open an archive browser that'll let you read the files inside without having to manually unzip the file first.
> 
> I'm going to go back to keeping that output in the initial details.

Thanks George, that worked for me. If this is documented on the commit workflow page on the wiki, I think it's fine, I was just struggling to find the build output like we had access to on Jenkins, and on GitHub was used to seeing the build output on the main page. If people test their changes locally first this shouldn't come up too often anyways.

**Review by jcolp — APPROVED** [review]

**Comment by InterLinked1** [comment]
No cherry-picks required

**Review by gtjoseph — APPROVED** [review]

**Pull request merged**